File: solutions/fsharp/largest-series-product/1/LargestSeriesProduct.fs
module LargestSeriesProduct

open System

let notDigit input =
    input |> Seq.forall Char.IsDigit |> not

let largestProduct (input: string) (seriesLength: int) : int option =
    if input.Length < seriesLength || seriesLength < 0 || notDigit input then
        None
    else
        List.init (input.Length - seriesLength + 1) (fun _ -> 1)
        |> List.mapi (fun i x ->
            input[i..i + seriesLength - 1]
            |> Seq.map (fun a -> int a - 48)
            |> Seq.reduce (fun y z -> y * z))
        |> List.max
        |> Some
        

File: solutions/scala/largest-series-product/1/src/main/scala/LargestSeriesProduct.scala
object Series {
  def largestProduct(span: Int, input: String): Option[Int] = {
    if (span <= 0 || span > input.length || input.exists(!_.isDigit)) {
      return None
    }

    val products = for {
      i <- 0 to input.length - span
      series = input.slice(i, i + span)
      product = series.map(_.asDigit).product
    } yield product

    if (products.isEmpty) None else Some(products.max)
  }
}
